Use inbound peerslot slots when a substream is received, rather than a connection

tomaka opened the pull request:
Fix #7074 

This implements the change described in #7074.

# New protocol between behaviour and handler

It took me a lot of time, but I came up with this refactoring of the communication between the `NetworkBehaviour` and the `ProtocolsHandler`. I believe that the new asynchronous protocol between the behaviour and handler is sound, in other words that it is not prone to race conditions.

The handler can now accept the following incoming messages:

- `Open`
- `Close`

It can also emit these messages:

- `OpenResultOk` or `OpenResultErr`.
- `CloseResult`.
- `OpenDesired`.
- `CloseDesired`.

Each `Open` message sent to the handler will result in a corresponding `OpenResultOk` or `OpenResultErr` being sent back. Each `Close` message will result in a corresponding `CloseResult` being sent back.
Thanks to this correspondence, the behaviour can track the state (open or closed) of the handler. For instance, if the behaviour sends an `Close`, then receives an `OpenDesired`, it knows that the `OpenDesired` was sent before the `Close` has been processed.

When in closed state, the handler can also send an `OpenDesired`, indicating that the remote would like the state to transition to "open".
When in open state, the handler can, vice-versa, send a `CloseDesired` indicating that the remote would like the state to transition to "closed".

As a small acceptable hack, an `OpenDesired` should always be answered by sending either `Open` or `Close`, otherwise the handler cannot differentiate between the situation where the behaviour is still undecided and the situation where the behaviour wants to remain in the "closed" state.
It is forbidden to remain "open" when a `CloseDesired` is received, and as such there is no similar ambiguity for that message. The behaviour must send `Close` as soon as possible.

# Inbound slots

At the moment, an inbound slot in the peerset is allocated whenever a TCP connection with a new peer is received.

This PR modifies this behaviour, and peerset inbound slots are now allocated when an `OpenDesired` is received from the handler.
Similarly, the slots are liberated when the substream is closed (this was already the case before).

This ties the inbound peerset slots to the concept of substreams, rather than the concept of TCP connections.

The motivation behind this change is to ultimately introduce peerset slots attributed to certain notification protocols (see https://github.com/paritytech/substrate/issues/6087). In the future, the messages emitted by the handler would be specific to a notifications protocol. For example, when a substream using the parachains collation protocol is opened, only a parachains collation slot is attributed.

# About `opening_and_closing`

In the implementation, I've opted to track, in the behaviour, the exact state of the handler. Not being tolerant will let us discover bugs sooner.

A consequence of tracking the exact state is that we verify that each `OpenResultOk`/`OpenResultErr`/`CloseResult` message emitted by the handler actually matches the state tracked in the behaviour.

But if the behaviour sends an `Open` message to the handler, then a `Close` message, then wants to send an `Open` message again, it starts being a lot of state to track. Rather than introducing a `Vec`, I went with putting the connection in a field called `opening_and_closing` and "banning" the peer for 5 seconds. In other words, we wait for 5 seconds before trying to send an `Open` message again.
Considering that this can only happen in case of quick state transitions coming from the peerset, I believe that this behaviour is completely ok, as it mimics the existing banning system.

tomaka commented:
Sorry for the lack of thorough explanations at the moment. This PR was very tiring to make, and I'm just very happy/relieved to finally open it.

tomaka commented:
Marked as ready for CI purposes only.

tomaka commented:
Ready for review.

tomaka commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -42,45 +42,34 @@ use wasm_timer::Instant;
+/// - Has open TCP connection(s), acknowledged by the peerset.
+///   - Notifications substreams are open on at least one connection, and external
+///     API has been notified.
+///   - Notifications substreams aren't open.
```
My intention is that these two bullets are "children" of the `Has open TCP connection(s), acknowledged by the peerset.` bullet.

In other words, it's actually two states:

- Has TCP connection and substreams open.
- Has TCP connection and substreams aren't open.

romanb replied:
Understood, I read these as two bullets applying _both_ to the parent bullet.

tomaka commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -94,22 +83,16 @@ use wasm_timer::Instant;
+///      first connection reports `NotifsHandlerOut::OpenResultOk`.
 ///   4. The behaviour reports `GenericProtoOut::CustomProtocolClosed` when the
-///      last connection reports `NotifsHandlerOut::Closed`.
+///      last connection reports `NotifsHandlerOut::ClosedResult`.
```
> as soon as the first connection reports closed substreams all connections would be closed and the closed state reported on the API

This is indeed the case before this PR.
The closing is however done indirectly: all the handler are switched to `Disabled`, which causes them to later report `Closed`, which closes the state.

Reverting this simply consists in *not* switching handlers to `Disabled`.

This PR restores the first/last scheme. However, in this PR, we also only try to enable only one connection at a time.
To clarify:

- If substreams are open on two different connections, and one of the connections closes, the state machine will transparently switch to the other (emitting a `CustomProtocolReplaced` event).
- If substreams are open on a connection, and a second connection opens, the state machine doesn't actively try to open the substreams of that second connection (contrary to before this PR).

romanb replied:
Understood and makes sense, thanks for the clarification.

tomaka commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -873,212 +1132,277 @@ impl NetworkBehaviour for GenericProto {
+					peer_id, *conn
+				);
+
+				if let Some(pos) = opening_and_closing.iter().position(|c| *c == *conn) {
+					opening_and_closing.remove(pos);
+				} else if let Some(pos) = closing.iter().position(|c| *c == *conn) {
+					closing.remove(pos);
+				} else if let Some(pos) = closed.iter().position(|c| *c == *conn) {
+					closed.remove(pos);
+				} else if let Some(pos) = open_desired.iter().position(|c| *c == *conn) {
```
I went for the approach of multiple fields because the list of fields isn't the same for all the variants of `PeerState`.
In particular, the `open_desired` field is only present in the `Incoming` state, and the `open` and `opening` fields are only present in `Enabled`.

romanb replied:
I understand, though I think it is a trade-off for the worse in terms of code complexity and possibilities for mistakes. Asserting that some connection states are never used by some peer states seems easier to me than making sure that this pop/push scheme between all these different vectors for state transitions is sound. I may be wrong, of course, as one usually only knows for sure after having actually tried it. Not to mention there is also the option of multiple state enums to precisely model the possible connection states in the different peer states.

tomaka replied:
I'll do the change, but after I've figured out why the tests are failing.

tomaka commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -873,212 +1132,277 @@ impl NetworkBehaviour for GenericProto {
+
+				if opening_and_closing.is_empty() && closed.is_empty() && closing.is_empty() {
+					if let Some(until) = banned_until {
+						*entry.get_mut() = PeerState::Banned { until };
```
That's a good point.
`Banned` concerns nodes that we aren't connected to, so if we removed them in `inject_disconnected`, that would be incorrect as well.

I'm going to fix that by adding a timer to them, and removing them in `poll` when the timer triggers.

tomaka commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -1127,197 +1453,434 @@ impl NetworkBehaviour for GenericProto {
 
-						*entry.into_mut() = PeerState::Disabled {
+						} else {
+							// Connections in `opening_and_closing` `opening` are in a Closed
```
```suggestion
							// Connections in `opening_and_closing` and `opening` are in a Closed
```

tomaka commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -1127,197 +1453,434 @@ impl NetworkBehaviour for GenericProto {
-								"State mismatch with {}: unknown closed connection",
-								source
-							);
+							// Connections in `opening_and_closing` `opening` are in a Closed
```
```suggestion
							// Connections in `opening_and_closing` and `opening` are in a Closed
```

tomaka commented:
I've found the issue: in the situation where the peer is in the `Enabled` state, and we have one connection `Open` and one `Closed`, and the `Open` connection closes, we don't try to open `Closed` and instead stay `Enabled` but with no opening will be ever be happening.

I'll fix that at the same time as I refactor the connections to be in only one list.

tomaka commented:
I did the refactoring. It looks like the tests are still failing, so I'll do more investigation/reviewing/debugging.

tomaka commented:
Since the new code no longer enables or disables the legacy substream at initialization, the timeout system of the legacy substream handler was being triggered, which would shut down connections after 20 seconds.

Rather than hacking around this system, I've pushed a commit that rewrites `legacy.rs` in order to take into account the current state of the protocol.

There is a debug assertion failing in `group.rs` which I haven't investigated yet.

tomaka commented:
After spending a lot of time trying to make `NotifsInHandler` and `NotifsOutHandler` work with the new design, I've finally decided to get rid of them and rewrite most of `group.rs` from scratch.

I'm mostly done, except that I've left some TODOs and commented out everything related to the legacy substream.
I've pushed a first commit now to check whether CI passes, and will finish the changes later.

tomaka commented:
The compilation warnings seem to prevent the CI from actually running, so I'll have to finish the PR before knowing.

tomaka commented:
I'm writing this comment before tests have passed, but I believe they do and that this should be ready for reviewing.
There isn't any more change coming from me except bugfixes.

tomaka commented:
This PR is currently being burned-in on `kusama-senty-ae1-0` with `debug-assertions = true`. It's been running since yesterday (Nov 12th) afternoon and hasn't panicked so far.

mxinden commented on client/network/src/protocol/generic_proto/handler.rs:
```diff
@@ -1,27 +1,1054 @@
+			message: message.into()
+		});
+
+		if result.is_err() {
```
In this case, the buffer is full, correct? Would one not want to disconnect from the node as describes in the function description?

tomaka replied:
This function is called from the very outer layers of `sc_network`, from `sc_network::NetworkService`.
The disconnect does happens when the handler receives the `NotificationsSinkMessage::ForceClose`.

mxinden replied:
Thanks for clarifying. I missed that the second `try_send` sends a `ForceClose` instead of the message itself.

mxinden commented on client/network/src/protocol/generic_proto/handler.rs:
```diff
@@ -1,27 +1,1054 @@
+				// Note: while we awknowledge legacy substreams and handle incoming messages,
+				// it doesn't trigger any `OpenDesired` event as a way to simplify the logic of
+				// this code.
+				// Since mid-2019, legacy substreams are supposed to used at the same time as
```
```suggestion
				// Since mid-2019, legacy substreams are supposed to be used at the same time as
```

mxinden commented on client/network/src/protocol/generic_proto/handler.rs:
```diff
@@ -1,27 +1,1054 @@
+			match substream.poll_next_unpin(cx) {
+				Poll::Ready(Some(Ok(_))) => {}
+				Poll::Pending => break,
+				Poll::Ready(Some(Err(_))) | Poll::Ready(None) => continue 'outer,
```
Would it not be helpful to log the substream error on `debug`?

tomaka replied:
This specific code handles errors on legacy substreams that we want to get rid of. As such, I don't think it's useful specifically here.

mxinden reviewed (commented):
Did a first pass over the changes. Thanks for documenting the code well! I am in favor of the many `debug_assert`.

tomaka commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -683,36 +800,50 @@ impl GenericProto {
-				error!(target: "sub-libp2p", "State of {:?} is poisoned", occ_entry.key()),
+			PeerState::Poisoned => {
+				error!(target: "sub-libp2p", "State of {:?} is poisoned", occ_entry.key());
+				debug_assert!(false);
```
That's debated. The coding rules that were established years ago mention that we should try to not panic as much as possible. In the `Poisoned` state, only a specific peer will be unable to connect to the node, as opposed to the entire node going down.

tomaka commented:
I believe I've addressed all concerns.
I'd appreciate a review of the actual details of all the state machine transitions, but I understand if you don't want to.

I suggest to continue the burnin (with `debug-assertions = true`) over the weekend, and merge early next week if nothing bad happened.

mxinden commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -638,42 +710,86 @@ impl GenericProto {
-				});
-				*occ_entry.into_mut() = PeerState::Enabled { open };
+			// Disabled => Enabled
+			PeerState::Disabled { mut connections, backoff_until } => {
```
```suggestion
			PeerState::Disabled { mut connections, backoff_until } => {
			debug_assert!(!connections.iter().any(|(_, s)| {
				matches!(s, ConnectionState::Open)
			}));			
```

Given that the back-off expired, one can assume that there are no open connections, right?

tomaka replied:
Even if the back-off hasn't expired, there should never be any connection open in the `PeerState::Disabled` state.

mxinden commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -1127,227 +1450,503 @@ impl NetworkBehaviour for GenericProto {
+						*entry.into_mut() = state;
+						return;
 					},
 					state => {
```
Do I understand correctly that it is not possible to receive a `NotifsHandlerOut::CloseDesired` for a `PeerState::Incoming`? Would it not be possible for a remote node to open and then close a connection before the local PSM has time to react?

tomaka replied:
In `PeerState::Incoming`, the handler is expected to be in the closed state, and as such receiving a `CloseDesired` would be a bug.

mxinden commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -1127,227 +1450,503 @@ impl NetworkBehaviour for GenericProto {
+				let mut entry = if let Entry::Occupied(entry) = self.peers.entry(source.clone()) {
+					entry
+				} else {
+					error!(target: "sub-libp2p", "OpenResultErr: State mismatch in the custom protos handler");
```
```suggestion
					error!(target: "sub-libp2p", "OpenResultErr: State mismatch in the custom protos handler");
					debug_assert!(false);
```

tomaka replied:
I had missed this one :confused:

mxinden reviewed (commented):
> But if the behaviour sends an Open message to the handler, then a Close message, then wants to send an Open message again, it starts being a lot of state to track. Rather than introducing a Vec, I went with putting the connection in a field called opening_and_closing and "banning" the peer for 5 seconds. In other words, we wait for 5 seconds before trying to send an Open message again.

This seems a bit like a hack to me, but to be honest I am unable to come up an alternative.

>  I'd appreciate a review of the actual details of all the state machine transitions, but I understand if you don't want to.

I took a deeper look resulting in the questions below. I agree that the mere size of these state machines makes it quite complex. I hope the `debug_assert`s bring up any outstanding bugs.

> I suggest to continue the burnin (with debug-assertions = true) over the weekend, and eventually merge early next week.

That sounds good to me.

mxinden commented on client/network/src/protocol/generic_proto/behaviour.rs:
```diff
@@ -260,6 +262,37 @@ impl PeerState {
+
+	/// Connection is in the `Closed` state, but a [`NotifsHandlerOut::OpenDesired`] message has
+	/// been received, meaning that the remote wants to open a substream.
+	OpenDesired,
```
> I believe I've addressed all concerns.

Friendly ping. I would suggest doing the same in behaviour as you have done for handler in https://github.com/paritytech/substrate/pull/7464/commits/ca31ed38de3a83c3e114d4d7e9a48d3767c87d22.

tomaka commented:
Since everything worked well during the weekend with `debug-assertions = true`, unless someone intends to review, I'm planning to merge this later today.

tomaka commented:
(this PR still needs at least one approving review to be merged)

tomaka commented:
bot merge

romanb reviewed (approved):
I still want to take a closer look at the most recent changes, but there is no reason to stall the PR. If I happen to have any important feedback, I will get in touch.

tomaka commented:
@romanb Thanks

To be clear, my eagerness to merge this PR is also due to the fact that #7374, which is the next thing I will work on, will require substantial modifications to `behaviour.rs` and `handler.rs` again.

tomaka commented:
bot merge
